cmd/skopeo/sync.go

```diff
@@ -0,0 +1,509 @@
+				"repo": repoName,
+			}).Info("Processing repo")
+
+			sourceCtx, err := contextFromGlobalOptions(c, "src-")
+			if err != nil {
+				logrus.WithFields(logrus.Fields{
+					"repo": repoName,
+				}).Error("Error getting source context, skipping")
+				logrus.Debug(err)
+				continue
+			}
+			// override ctx with per-server options
+			sourceCtx.DockerCertPath = serverCfg.CertDir
+			sourceCtx.DockerDaemonCertPath = serverCfg.CertDir
+			sourceCtx.DockerDaemonInsecureSkipTLSVerify = !serverCfg.TLSVerify
+			sourceCtx.DockerInsecureSkipTLSVerify = !serverCfg.TLSVerify
+			sourceCtx.DockerAuthConfig = &serverCfg.Credentials
+
+			var sourceReferences []types.ImageReference
+			for _, tag := range tags {
+				source := fmt.Sprintf("%s:%s", repoName, tag)
+
+				imageRef, err := getImageReference(source)
+				if err != nil {
+					logrus.WithFields(logrus.Fields{
+						"tag": source,
+					}).Error("Error processing tag, skipping")
+					logrus.Debugf("Error getting image reference: %s", err)
+					continue
+				}
+				sourceReferences = append(sourceReferences, imageRef)
+			}
+
+			if len(tags) == 0 {
+				imageRef, err := getImageReference(repoName)
+				if err != nil {
+					logrus.WithFields(logrus.Fields{
+						"repo": repoName,
+					}).Error("Error processing repo, skipping")
+					logrus.Debug(err)
+					continue
+				}
+				logrus.WithFields(logrus.Fields{
+					"repo": repoName,
+				}).Info("No tags given, querying registry")
+
+				sourceReferences, err = imagesToCopyFromRegistry(imageRef, repoName, sourceCtx)
+				if err != nil {
+					logrus.WithFields(logrus.Fields{
+						"repo": repoName,
+					}).Error("Error processing repo, skipping")
+					logrus.Debug(err)
+					continue
+				}
+			}
+
+			repoDescList = append(repoDescList, repoDescriptor{
+				RootReference:  nil,
+				TagsReferences: sourceReferences,
+				Context:        sourceCtx})
+		}
+	}
+
+	return
+}
+
+func syncHandler(c *cli.Context) (retErr error) {
+	if len(c.Args()) != 1 {
+		cli.ShowCommandHelp(c, "sync")
+		return errors.New("Exactly one argument expected")
+	}
+
+	if c.IsSet("source") && c.IsSet("source-file") {
+		return fmt.Errorf(
+			"Cannot use the '--source' and '--source-file' flags at the same time")
+	}
+
+	if !c.IsSet("source") && !c.IsSet("source-file") {
+		return fmt.Errorf(
+			"Must specify either a '--source' or a '--source-file'")
+	}
+
+	policyContext, err := getPolicyContext(c)
+	if err != nil {
+		return fmt.Errorf("Error loading trust policy: %v", err)
+	}
+	defer policyContext.Destroy()
+
+	destRef, err := getImageReference(c.Args()[0])
+	if err != nil {
+		return fmt.Errorf("Error while parsing destination: %v", err)
+	}
+	destinationCtx, err := contextFromGlobalOptions(c, "dest-")
+	if err != nil {
+		return err
+	}
+
+	signBy := c.String("sign-by")
+	removeSignatures := c.Bool("remove-signatures")
+
+	//TODO: should we assume that's our default manifest type?
+	manifestType := manifest.DockerV2Schema2MediaType
```

Comment:
I think so, but I'd prefer a way to pass in a different variant.  But I'd lean on @mtrmac or @runcom for the final word.

Reply:
My thinking was just to make the code more flexible if ever needed.  @mtrmac if you think that's not necessary or the cost isn't worth it, I'm fine with that.  Just preserving the manifest as you suggested makes sense to me too.

Reply:
I’m _weakly_ against exposing options just because the underlying code implements them: it takes a possibly unnecessary effort, and after the options are introduced, it’s difficult to both remove them and to gauge whether they have been ever necessary in the first place.

But in this case, I’m _primarily_ saying that the default behavior should, unless there is a specific reason, not be _hard-coding_ manifest types at all. The existing autodetection is expected to just work, and it is future-proof, unlike hard-coding a specific default which is certain to become obsolete at some point.